### PR TITLE
SYS-3560 Adds event listener feature to the node

### PR DESCRIPTION
**Pull request by thadouk**
## Proposed changes
Extends Ethereum bridge pallet to be bidirectional and listen to events emitted in the bridge contract.
The PR introduces:

- `avn-event-handler` service: a thread running in client space that acts as an oracle, discovering information from Ethereum and submitting the information back to the runtime
- The information is submitted in batches based on an Ethereum block range which is agreed initially from collators and then dictated by the runtime
- New mechanism is configured to co-exist with ethereum-events and avoid double processing of events in the system.

## Type of change/Merge

🚨What type of change is this PR? </br>
_Put an `x` in the boxes that apply_

- [x] Release 
  - [ ] Increase versions 
  - [ ] Baseline tests passed  
  - Release type:
    - [ ] Major release 
    - [x] Minor release 
    - [ ] Patch release 

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._

- [ ] You describe the purpose of the PR, e.g.:
  - What does it do?
  - Highlight what important points reviewers should know about;
  - Indicates if there is something left for follow-up PRs.
- [ ] Documentation updated
- [ ] Business logic tested successfully
- [ ] Verify First, Write Last: In Substrate development, it is important that you always ensure preconditions are met and return errors at the beginning. After these checks have completed, then you may begin the function's computation.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,7 @@ members = [
     "node/avn-service",
     "node/avn-lower-rpc",
     "pallets/*",
+  	"pallets/*/runtime-api",
```
Comment (ivan-cholakov) [inline Cargo.toml:29]: you see the indentation bug, it is happening to you too :D

File: Dockerfile.22_04 [hunk 2]
```diff
@@ -24,6 +24,7 @@ RUN apt-get update && \
 	ca-certificates \
 	curl \
 	jq && \
+	update-ca-certificates && \
```
Comment (ivan-cholakov) [inline Dockerfile.22_04:27]: :100:

**Review by ivan-cholakov — APPROVED** [review]

File: node/src/command.rs [hunk 3]
```diff
@@ -233,22 +233,30 @@ pub fn run() -> Result<()> {
...
-
-            runner.async_run(|_| {
-                Ok((cmd.run::<Block, HostFunctionsOf<ParachainExecutor>>(), task_manager))
+            use try_runtime_cli::block_building_info::substrate_info;
+
+            let runner = cli.create_runner(cmd)?;
+            runner.async_run(|config| {
+                // we don't need any of the components of new_partial, just a runtime, or a task
+                // manager to do `async_run`.
+                let registry = config.prometheus_config.as_ref().map(|cfg| &cfg.registry);
```
Comment (ivan-cholakov) [inline node/src/command.rs:246]: you now create the `task_manager` inside the runner? why is this change needed?

Comment (nahuseyoum) [reply node/src/command.rs:246]: it doesn't compile with the previous code when you enable try-runtime. This is copied from Substrate

File: pallets/eth-bridge/src/lib.rs [hunk 4]
```diff
@@ -485,6 +563,140 @@ pub mod pallet {
             Self::deposit_event(Event::<T>::ActiveRequestRemoved { request_id });
             Ok(().into())
         }
+
+        // TODO update weights
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:567]:
```suggestion
```

File: pallets/eth-bridge/src/lib.rs [hunk 5]
```diff
@@ -485,6 +563,140 @@ pub mod pallet {
...
+                <T as Config>::WeightInfo::submit_ethereum_events_and_process_batch(
+                    MAX_VALIDATOR_ACCOUNTS,
+                    MAX_INCOMING_EVENTS_BATCH_SIZE,
+                )
+            };
+
+            Ok(Some(final_weight).into())
+        }
+
+        // TODO update weights
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:619]:
```suggestion
```

File: pallets/eth-bridge/src/lib.rs [hunk 6]
```diff
@@ -604,6 +816,110 @@ pub mod pallet {
...
+                        "Unknown Ethereum event signature in range {:?}",
+                        &discovered_event.event.event_id.signature
+                    );
+                },
+            }
+        }
+
+        // Cleanup
+        for (partition, votes) in EthereumEvents::<T>::drain() {
+            // TODO raise offences
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:891]: do we have a ticket for this?

File: pallets/eth-bridge/src/tests/incoming_events_tests.rs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+};
+use sp_runtime::{testing::UintAuthorityId, DispatchError};
+pub extern crate alloc;
+use alloc::collections::BTreeSet;
+use sp_avn_common::{event_discovery::EthBridgeEventsFilter, event_types::ValidEvents};
+
+const ROOT_HASH: &str = "30b83f0d722d1d4308ab4660a72dbaf0a7392d5674eca3cd21d57256d42df7a0";
+const REWARDS: &[u8] = b"15043665996000000000";
+const AVG_STAKED: &[u8] = b"9034532443555111110000";
+const PERIOD: &[u8] = b"3";
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/tests/incoming_events_tests.rs:16]: Not sure if these are used anywhere in this test. Maybe we can tidy them up in a separate PR.

File: pallets/token-manager/src/lib.rs [hunk 8]
```diff
@@ -1092,4 +1096,8 @@ impl<T: Config> BridgeInterfaceNotification for Pallet<T> {
 
         Ok(())
     }
+
+    fn on_incoming_event_processed(event: &EthEvent) -> DispatchResult {
+        Self::processed_event_handler(event)
```
Comment (nahuseyoum) [inline pallets/token-manager/src/lib.rs:1101]: Is it worth adding the feature guard here for `Lift` and `Growth` coming via this hook?

File: primitives/avn-common/src/event_discovery.rs [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+impl EthBlockRange {
+    pub fn next_range(&self) -> EthBlockRange {
+        EthBlockRange {
+            start_block: self.start_block.saturating_add(self.length),
+            length: self.length,
+        }
+    }
+    pub fn range(&self) -> (u32, u32) {
+        let end_block = self.start_block.saturating_add(self.length).saturating_less_one();
```
Comment (nahuseyoum) [inline primitives/avn-common/src/event_discovery.rs:29]: Why do we remove 1? Not sure if the tests mirror this.

File: primitives/avn-common/src/event_discovery.rs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    );
+    (context, &account_id, data).encode()
+}
+
+pub mod events_helpers {
+    use super::*;
+    pub extern crate alloc;
+    use alloc::collections::BTreeSet;
+
+    pub fn discovered_eth_events_partition_factory(
```
Comment (nahuseyoum) [inline primitives/avn-common/src/event_discovery.rs:147]: Nit: this doesn't feel like a function name :) maybe it should start with a verb

File: primitives/avn-common/src/event_discovery.rs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        }
+        partitions
+    }
+
+    // TODO unit test this
+    pub fn compute_finalised_block_number(
+        ethereum_block: u32,
+        range_length: u32,
+    ) -> Result<u32, ()> {
+        let calculation_block = ethereum_block.saturating_sub(5 * range_length);
```
Comment (nahuseyoum) [inline primitives/avn-common/src/event_discovery.rs:194]: :+1:

File: runtime/avn/src/lib.rs [hunk 12]
```diff
@@ -936,6 +964,59 @@ impl_runtime_apis! {
...
+        }
+
+        fn query_active_block_range()-> Option<(EthBlockRange, u16)> {
+            if let Some(active_eth_range) =  EthBridge::active_ethereum_range(){
+                Some((active_eth_range.range, active_eth_range.partition))
+            } else {
+                None
+            }
+        }
+        fn query_has_author_casted_vote(account_id: AccountId) -> bool{
```
Comment (nahuseyoum) [inline runtime/avn/src/lib.rs:990]:
```suggestion
        
        fn query_has_author_casted_vote(account_id: AccountId) -> bool{
```

**Review by nahuseyoum — APPROVED** [review]

File: primitives/avn-common/src/event_discovery.rs (hunk 9 above)
Comment (thadouk) [reply primitives/avn-common/src/event_discovery.rs:29]: EthBlockRange should be `[start_block, end_block)` where `end_block = start_block + length`

Comment (nahuseyoum) [reply primitives/avn-common/src/event_discovery.rs:29]: Exactly but if am reading the code correctly, we are doing `end_block = start_block + length - 1`

Comment (thadouk) [reply primitives/avn-common/src/event_discovery.rs:29]: so the range is exclusive of the last block 
so for an instance of EthereumRange { start:block: 0 , length: 100 }

range 1 is : mathematically `[0,100)` or `[0,99]`. Rust wise is `0..=99` or `0..100`
range 2 is : mathematically `[100,200)` or `[100,199]`. Rust wise is `100..=199` or `100..200`

I think the formula does that

Comment (thadouk) [reply primitives/avn-common/src/event_discovery.rs:29]: But since we are talking about this its worth validating that when we query the range returned in the client, the full range (including the range start and end)

File: pallets/eth-bridge/src/lib.rs (hunk 6 above)
Comment (thadouk) [reply pallets/eth-bridge/src/lib.rs:889]: we do now

File: pallets/token-manager/src/lib.rs (hunk 8 above)
Comment (thadouk) [reply pallets/token-manager/src/lib.rs:1101]: No.
from the pallet point of view we either process events coming from eth-bridge or not.
The runtime defines the filter and the rest of the components are just executing.

File: primitives/avn-common/src/event_discovery.rs (hunk 10 above)
Comment (thadouk) [reply primitives/avn-common/src/event_discovery.rs:147]: typically you don't use verbs on a factory method as it doesn't make sense. I can move it under a struct that is the factory
